packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -276,3 +276,258 @@ class CustomUserData extends UserData {
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  readonly transferEncoding?: string;
+}
+
+/**
+ * Options when creating `MultipartUserDataPart`.
+ */
+export interface MultipartUserDataPartOptionsWithBody extends MultipartUserDataPartOptions {
+  /**
+   * The body of message.
+   *
+   * @default undefined - body will not be added to part
+   */
+  readonly body?: string,
+}
+
+/**
+ * Options when creating `MultipartUserDataPartWrapper`.
+ */
+export interface MultipartUserDataPartWrapperOptions {
+  /**
+   * `Content-Type` header of this part.
+   *
+   * For Linux shell scripts typically it's `text/x-shellscript`.
+   *
+   * @default 'text/x-shellscript; charset="utf-8"'
+   */
+  readonly contentType?: MultipartContentType;
+}
+
+/**
+ * Interface representing part of `MultipartUserData` user data.
+ */
+export interface IMultipart {
+  /**
+   * The body of this MIME part.
+   */
+  readonly body: string | undefined;
+
+  /**
+   * `Content-Type` header of this part.
+   */
+  readonly contentType: string;
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  readonly transferEncoding?: string;
+}
+
+/**
+ * The base class for all classes which can be used as {@link MultipartUserData}.
+ */
+export abstract class MultipartUserDataPart implements IMultipart {
+
+  /**
+   * Constructs the new `MultipartUserDataPart` wrapping existing `UserData`. Modification to `UserData` are reflected
+   * in subsequent renders of the part.
+   *
+   * For more information about content types see `MultipartUserDataPartOptionsWithBody`
+   */
+  public static fromUserData(userData: UserData, opts?: MultipartUserDataPartWrapperOptions): MultipartUserDataPart {
+    opts = opts || {};
+    return new MultipartUserDataPartWrapper(userData, opts);
+  }
+
+  /**
+   * Constructs the raw `MultipartUserDataPart` using specified body, content type and transfer encoding.
+   *
+   * When transfer encoding is specified (typically as Base64), it's caller responsibility to convert body to
+   * Base64 either by wrapping with `Fn.base64` or by converting it by other converters.
+   */
+  public static fromRawBody(opts: MultipartUserDataPartOptionsWithBody): MultipartUserDataPart {
+    return new MultipartUserDataPartRaw(opts);
+  }
+
+  protected static readonly DEFAULT_CONTENT_TYPE = 'text/x-shellscript; charset="utf-8"';
+
+  /** The body of this MIME part. */
+  public abstract get body(): string | undefined;
+
+  /** `Content-Type` header of this part */
+  public readonly contentType: string;
+
+  /**
+   * `Content-Transfer-Encoding` header specifying part encoding.
+   *
+   * @default undefined - don't add this header
+   */
+  public readonly transferEncoding?: string;
+
+  public constructor(props: MultipartUserDataPartOptions) {
+    this.contentType = props.contentType;
```

Comment:
We define this class for its interface. Let's not muddy it up with implementation inheritance either.

I'd rather have the constructor be empty and the fields you set here be abstract, to be filled in the children. With this constructor, we won't be able to change the behavior of these getters (for example, make them lazy or something).

Reply:
I removed `IMultipart` interface and moved the part render method here